### PR TITLE
Reader: shuffle row groups before sharding.

### PR DESCRIPTION
**Pull request by chongxiaoc**
Splitting row groups by % num_shards is not a good idea to split data.
This commit introduce a shard_seed parameter to random shuffle row
groups before sharding.

### CODE REVIEW

File: petastorm/reader.py [hunk 1]
```diff
@@ -378,8 +378,10 @@ def __init__(self, pyarrow_filesystem, dataset_path, schema_fields=None,
         :param filters: (List[Tuple] or List[List[Tuple]]): Standard PyArrow filters.
             These will be applied when loading the parquet file with PyArrow. More information
             here: https://arrow.apache.org/docs/python/generated/pyarrow.parquet.ParquetDataset.html
+        :param shard_seed: Random seed to shuffle row groups for data sharding. Defaults to None
         """
         self.num_epochs = num_epochs
+        self.shard_seed = shard_seed
```
Comment (selitvin) [inline petastorm/reader.py:384]: Since we use this value only from the constructor , suggest we don't turn it into a state variable, but pass it down the call-stack to the consumer. By limiting the scope of the variable, it makes the code more readable, IMO.

**Review by selitvin — CHANGES_REQUESTED** [review]
Please add a unit test to cover the new case.

File: petastorm/reader.py (hunk 1 above)
Comment (chongxiaoc) [reply petastorm/reader.py:384]: sounds reasonable to me.

**Comment by chongxiaoc** [comment]
@selitvin I addressed the comments, can you take one more look?

File: petastorm/reader.py [hunk 2]
```diff
@@ -550,6 +556,13 @@ def _partition_row_groups(self, dataset, row_groups, shard_count, cur_shard,
 
         # We hash on the relative path of each parquet file to guarantee consistency between different reader
         # constructions even after moving the dataset
+        if shard_seed:
+            import random
```
Comment (selitvin) [inline petastorm/reader.py:560]: Do you know why these lines appear as not covered? Is anything wrong with either the test or the code?

Comment (chongxiaoc) [reply petastorm/reader.py:560]: hmmm, I modified one unit test to cover `shard_seed` != None case. Let me double check.

Comment (chongxiaoc) [reply petastorm/reader.py:560]: Should be fixed. I used `0` for shard_seed, which make `if shard_seed` false.

**Comment by chongxiaoc** [comment]
@selitvin please take a look again, fixed codecov issues.

**Review by selitvin — APPROVED** [review]
Sorry, forgot to ask for one last thing. Please update the release notes (docs/release-notes.rst).

**Comment by chongxiaoc** [comment]
@selitvin updated release notes
